scipy/sparse/csgraph/_matching.pyx

```diff
@@ -0,0 +1,276 @@
+    directly as follows:
+
+    >>> data = [0, 0, 0]
+    >>> indices = [2, 0, 1]
+    >>> indptr = [0, 1, 3]
+    >>> graph = csr_matrix((data, indices, indptr))
+    >>> print(maximum_bipartite_matching(graph, perm_type='column'))
+    [2 0]
+    >>> print(maximum_bipartite_matching(graph, perm_type='row'))
+    [ 1 -1  0]
+
+    When one or both of the partitions are empty, the matching is empty as
+    well:
+
+    >>> graph = csr_matrix((2, 0))
+    >>> print(maximum_bipartite_matching(graph, perm_type='column'))
+    [-1 -1]
+    >>> print(maximum_bipartite_matching(graph, perm_type='row'))
+    []
+
+    When the input matrix is square, and the graph is known to admit a perfect
+    matching, i.e. a matching with the property that every vertex in the graph
+    belongs to some edge in the matching, then one can view the output as the
+    permutation of rows (or columns) turning the input matrix into one with the
+    property that all diagonal elements are non-empty:
+
+    >>> a = [[0, 1, 2, 0], [1, 0, 0, 1], [2, 0, 0, 3], [0, 1, 3, 0]]
+    >>> graph = csr_matrix(a)
+    >>> perm = maximum_bipartite_matching(graph, perm_type='row')
+    >>> print(graph[perm].toarray())
+    [[1 0 0 1]
+     [0 1 2 0]
+     [0 1 3 0]
+     [2 0 0 3]]
+
+    """
+    if isspmatrix_csc(graph) or isspmatrix_coo(graph):
+        graph = graph.tocsr()
+    elif not isspmatrix_csr(graph):
+        raise TypeError("graph must be in CSC, CSR, or COO format.")
+    i, j = graph.shape
+    x, y = _hopcroft_karp(graph.indices, graph.indptr, i, j)
+    return np.asarray(x if perm_type == 'column' else y)
+
+
+@cython.boundscheck(False)
+@cython.wraparound(False)
+cdef tuple _hopcroft_karp(ITYPE_t[:] indices, ITYPE_t[:] indptr,
+                          ITYPE_t i, ITYPE_t j):
+    cdef ITYPE_t INF = np.iinfo(ITYPE).max
+    # x will end up containing the matchings of rows to columns, while
+    # y will contain the matchings of columns to rows.
+    cdef ITYPE_t[:] x = np.empty(i, dtype=ITYPE)
+    cdef ITYPE_t[:] y = np.empty(j, dtype=ITYPE)
+
+    # During the BFS step, dist will keep track of the level of the search. We
+    # only keep track of this for the vertices in the left partition. Note that
+    # the array is one longer than the cardinality of the left partition:
+    # throughout the algorithm, we make use of the standard trick of adding an
+    # auxiliary vertex whose index will be i, and whose semantics are that
+    # every unmatched column will be matched with this vertex.
+    cdef ITYPE_t[:] dist = np.empty(i + 1, dtype=ITYPE)
+
+    cdef ITYPE_t k, v, w, up, u, u_old
+
+    # At the end of the day, unmatched vertices will have a value of -1. As
+    # mentioned above, unmatched vertices in the right partition will be
+    # matched to the auxiliary vertex i, and we will set their value to -1
+    # only when we have found a maximum matching.
+    for k in range(i):
+        x[k] = -1
+
+    for k in range(j):
+        y[k] = i
```

Comment:
perhaps simpler than these for-loops:
```
x.fill(-1)
y.fill(i)
```

Reply:
Makes sense. I did it the other way to ensure that, outside of array initialization, the method had no Python calls at all, but since `fill` is fast(er?) and cleaner, and this is outside of the hot loop, let me just include your suggestion.

Reply:
Looks like this won't quite work without changing the types as well:

```python
AttributeError: 'scipy.sparse.csgraph._matching._memoryviewslice' object has no attribute 'fill'
```

Reply:
Ah sorry about that, I didn't test my proposed change. `memoryview` is mostly compatible with `ndarray`, but not fully - I was just looking through numpy-colored glasses.